src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,668 @@
+   *     hh:mm:ss   : ^(GMT|UTC)?[+-](\d|0[0-9]|1[0-8]):[0-5][0-9]:[0-5][0-9]
+   *   without colon
+   *     hh only    : ^(GMT|UTC)?[+-](\d|0[0-9]|1[0-8])
+   *     hh:mm:(ss) : ^(GMT|UTC)?[+-](0[0-9]|1[0-8])([0-5][0-9])?([0-5][0-9])?
+   *
+   *   additional restriction: 18:00:00 is the upper bound (which means 18:00:01
+   * is invalid)
+   */
+  __device__ inline thrust::pair<int64_t, uint8_t> parse_utc_like_tz(
+    string_view const& tz_lit) const
+  {
+    size_type len = tz_lit.size_bytes();
+
+    char const* ptr = tz_lit.data();
+
+    size_t char_offset = 0;
+    // skip UTC|GMT if existing
+    if (len > 2 && ((*ptr == 'G' && *(ptr + 1) == 'M' && *(ptr + 2) == 'T') ||
+                    (*ptr == 'U' && *(ptr + 1) == 'T' && *(ptr + 2) == 'C'))) {
+      char_offset = 3;
+    }
+
+    // return for the pattern UTC|GMT (without exact offset)
+    if (len == char_offset) return {0, 0};
+
+    // parse sign +|-
+    char const sign_char = *(ptr + char_offset++);
+    int64_t sign;
+    if (sign_char == '+') {
+      sign = 1L;
+    } else if (sign_char == '-') {
+      sign = -1L;
+    } else {
+      // if the rep starts with UTC|GMT, it can NOT be region-based rep
+      return {0, char_offset < 3 ? 1 : 2};
+    }
+
+    // parse hh:mm:ss
+    int64_t hms[3] = {0L, 0L, 0L};
+    bool has_colon = false;
+    for (size_type i = 0; i < 3; i++) {
+      // deal with the first digit
+      hms[i] = *(ptr + char_offset++) - '0';
+      if (hms[i] < 0 || hms[i] > 9) return {0, 2};
+
+      // deal with trailing single digit instant:
+      //  hh(GMT+8) - valid
+      //  mm(GMT+11:2) - must be separated from (h)h by `:`
+      //  ss(GMT-11:22:3) - invalid
+      if (len == char_offset) {
+        if (i == 2 || (i == 1 && !has_colon)) return {0, 2};
+        break;
+      }
+
+      // deal with `:`
+      if (*(ptr + char_offset) == ':') {
+        // 1. (i == 1) one_digit mm with ss is invalid (+11:2:3)
+        // 2. (i == 2) one_digit ss is invalid (+11:22:3)
+        // 3. trailing `:` is invalid (GMT+8:)
+        if (i > 0 || len == ++char_offset) return {0, 2};
+        has_colon = true;
+        continue;
+      }
+
+      // deal with the second digit
+      auto digit = *(ptr + char_offset++) - '0';
+      if (digit < 0 || digit > 9) return {0, 2};
+      hms[i] = hms[i] * 10 + digit;
+
+      if (len == char_offset) break;
+      // deal with `:`
+      if (*(ptr + char_offset) == ':') {
+        // trailing `:` is invalid (UTC+11:)
+        if (len == ++char_offset) return {0, 2};
+        has_colon = true;
+      }
+    }
+
+    // the upper bound is 18:00:00 (regardless of sign)
+    if (hms[0] > 18 || hms[1] > 59 || hms[2] > 59) return {0, 2};
+    if (hms[0] == 18 && hms[1] + hms[2] > 0) return {0, 2};
+
+    return {sign * (hms[0] * 3600L + hms[1] * 60L + hms[2]), 0};
+  }
+
+  /**
+   * TODO: replace linear search with more efficient approach (like prefix tree)
```

Comment:
Is there an issue for this? Should we just sort the list alphabetically on the CPU before sending it down and then do a binary search?

Reply:
Done.